python/ray/data/preprocessor.py

```
@@ -216,69 +216,41 @@ def _fit(self, dataset: "Dataset") -> "Preprocessor":
+        * If only `_transform_pandas` is implemented, then use ``pandas`` batch format.
+        * If only `_transform_numpy` is implemented, then use ``numpy`` batch format.
+        * If both are implemented, then use the Preprocessor defined preferred batch
+        format.
         """
 
-        assert data_format in (
-            "pandas",
-            "arrow",
-            "numpy",
-        ), f"Unsupported data format: {data_format}"
-
         has_transform_pandas = (
             self.__class__._transform_pandas != Preprocessor._transform_pandas
         )
         has_transform_numpy = (
             self.__class__._transform_numpy != Preprocessor._transform_numpy
         )
 
-        # Infer transform type by prioritizing native transformation to minimize
-        # data conversion cost.
-        if data_format == BlockFormat.PANDAS:
-            # Perform native pandas transformation if possible.
-            if has_transform_pandas:
-                transform_type = BatchFormat.PANDAS
-            elif has_transform_numpy:
-                transform_type = BatchFormat.NUMPY
-            else:
-                raise NotImplementedError(
-                    "None of `_transform_numpy` or `_transform_pandas` "
-                    f"are implemented for dataset format `{data_format}`."
-                )
-        elif data_format == BlockFormat.ARROW or data_format == "numpy":
-            # Arrow -> Numpy is more efficient
-            if has_transform_numpy:
-                transform_type = BatchFormat.NUMPY
-            elif has_transform_pandas:
-                transform_type = BatchFormat.PANDAS
-            else:
-                raise NotImplementedError(
-                    "None of `_transform_numpy` or `_transform_pandas` "
-                    f"are implemented for dataset format `{data_format}`."
-                )
-
-        return transform_type
+        if has_transform_numpy and has_transform_pandas:
```

I wonder if we should just raise an error in this case? It's kind of weird the preprocessor should have both implemented. If we raised an error, we could also avoid needing this new preferred_batch_format API.